[SUBGRAPH/SDK-CORE] | Issues #1273 - ATS properties, 1291 - configs on Token, 1292 - userData on Stream

0xdavinchee opened the pull request:
## Summary
#1273 - New aggregate properties added for ATS and ATSLog entities:
- `activeOutgoingStreamCount`
- `activeIncomingStreamCount`
- `inactiveOutgoingStreamCount`
- `inactiveIncomingStreamCount`

#1291 - Governance Configs on `Token` entity
- `rewardAddress`, `liquidationPeriod`, `patricianPeriod`, and `minimumDeposit` added to `Token`
- New `DefaultGovernanceConfigs` HOL entity created for single entity containing our default governance parameters for all tokens if they don't have a config set specifically (when the `superToken` field is 0x0)
- Added `SuperTokenMinimumDepositChangedEvent` entity which was previously missing

cc @vmichalik 

#1292 - `userData` field added to `Stream` entity
- `userData` property added to `Stream` and represents the last updated `userData`

Misc
- @kasparkallas requested to throw an error on BatchCall in SDK-Core as well

SDK-Core subgraph mapping added as well.

0xdavinchee committed:
add userData to Stream entity

0xdavinchee committed:
governance configs

- create SuperTokenMinimumDepositChangedEvent event entity
- create DefaultGovernanceConfigs HOL entity
- add config fields to Token entity

0xdavinchee committed:
new aggregate properties on ATS

- activeOutgoingStreamCount, activeIncomingStreamCount, inactiveOutgoingStreamCount, inactiveIncomingStreamCount

0xdavinchee committed:
SuperTokenMinimumDepositChanged for SDK-Core

- added SuperTokenMinimumDepositChanged for SDK-Core
- files needed to be changed:
  - src/mapGetAllEventsQueryEvents.ts
  - src/subgraph/events/events.graphql
  - src/subgraph/queries/getAllEvents.graphql
  - regenerate src/subgraph/schema.graphql with locally deployed subgraph

0xdavinchee committed:
added hol entity properties

- account token snapshot, stream and token
- edit query files in
src/subgraph/entities AND src/subgraph/queries

0xdavinchee committed:
add ats properties to ats log

0xdavinchee committed:
add properties in SDK core for ats log

0xdavinchee committed:
handle assumption

0xdavinchee committed:
fixes

- compiler error w/ assemblyscript
- forgot to regenerate schema.graphql after adding atslog properties

0xdavinchee committed:
throw on empty batch call

0xdavinchee committed:
Update CHANGELOG.md

0xdavinchee committed:
Merge branch 'dev' into subgraph-user-data

0xdavinchee committed:
fix deploy-subgraph.yml

kasparkallas commented on packages/subgraph/src/mappingHelpers.ts:
```diff
@@ -143,6 +163,30 @@ export function getOrInitSuperToken(
+        token.patricianPeriod = defaultGovernanceConfigs.patricianPeriod;
+    }
+
+    if (token.minimumDeposit.equals(BIG_INT_ZERO) && defaultGovernanceConfigs != null) {
```
Nest all the conditions in a single `if (defaultGovernanceConfigs != null)` to repeat less?

0xdavinchee replied:
agree and resolved

kasparkallas commented on packages/subgraph/src/mappingHelpers.ts:
```diff
@@ -154,6 +198,34 @@ export function getOrInitSuperToken(
+    );
+
+    if (defaultGovernanceConfigs == null) {
+        defaultGovernanceConfigs = new DefaultGovernanceConfigs(
```
Rename to singular? "DefaultGovernanceConfig"

0xdavinchee replied:
agree and resolved

kasparkallas commented on packages/subgraph/src/mappingHelpers.ts:
```diff
@@ -154,6 +198,34 @@ export function getOrInitSuperToken(
+        defaultGovernanceConfigs.createdAtTimestamp = block.timestamp;
+        defaultGovernanceConfigs.createdAtBlockNumber = block.number;
+        defaultGovernanceConfigs.rewardAddress = ZERO_ADDRESS;
+        defaultGovernanceConfigs.liquidationPeriod = BIG_INT_ZERO;
```
What's the point of the default config if even `liquidationPeriod` is zero?

0xdavinchee replied:
because it is the true "default", default in this case refers to what is set for a config (e.g. rewardAddress, liqPeriod, patPeriod, minDeposit) where the superToken field is `address(0)`.
each of these fields are only set at the protocol (and therefore the subgraph) when an event is emitted indicating that the default value has been set for a particular config.
it is better to initialize the entity with properties = 0 which means nothing has been set yet than to initialize at a value (which may be incorrect) we've used in the past.

kasparkallas commented on packages/subgraph/src/mappings/superfluidGovernance.ts:
```diff
@@ -41,6 +46,29 @@ export function handleRewardAddressChanged(event: RewardAddressChanged): void {
     ev.save();
 
+    // Update reward address for token or default governance configs
+    if (event.params.superToken.equals(ZERO_ADDRESS)) {
```
Why would this ever be 0-address? Is this a pattern in the protocol?

0xdavinchee replied:
yes, see the comment above or:

default configs refers to configs set for `event.params.superToken = address(0)`, that is, the super token parameter/argument for a governance config is address(0).

kasparkallas commented on packages/subgraph/src/mappings/superfluidGovernance.ts:
```diff
@@ -41,6 +46,29 @@ export function handleRewardAddressChanged(event: RewardAddressChanged): void {
+            ZERO_ADDRESS.toHexString()
+        );
+        if (defaultGovernanceConfigs) {
+            defaultGovernanceConfigs.rewardAddress = event.params.rewardAddress;
```
I don't think `updated*` fields are getting updated on this code path.

0xdavinchee replied:
I am assuming you are referring to governance configs being updated here and that case is handled here.

kasparkallas replied:
I mean I don't notice these getting updated:
<img width="789" alt="image" src="https://user-images.githubusercontent.com/10894666/223146550-2884ed8a-660f-4e78-bd6e-c468c80fb6da.png">

kasparkallas commented on packages/subgraph/schema.graphql:
```diff
@@ -1656,6 +1688,43 @@ type StreamPeriod @entity {
+DefaultGovernanceConfigs: A higher order entity that represents the default governance configs.
+When a config is set in governance with superToken field as address(0), it will be used as the default config.
+"""
+type DefaultGovernanceConfigs @entity {
```
So there is this one default configuration entity? And when there's anything token specific then that is saved on the token?

0xdavinchee replied:
yes

0xdavinchee committed:
Update prepare-manifest.sh

0xdavinchee committed:
address review comments

- Nest all the conditions in a single if (defaultGovernanceConfigs != null) to repeat less
- Rename to singular "DefaultGovernanceConfig"

kasparkallas commented on packages/subgraph/src/mappings/superfluidGovernance.ts:
```diff
@@ -61,6 +89,31 @@ export function handleCFAv1LiquidationPeriodChanged(
+
+    // Update cfav1 liquidation period for token or default governance configs
+    if (event.params.superToken.equals(ZERO_ADDRESS)) {
+        let defaultGovernanceConfigs = DefaultGovernanceConfigs.load(
```
Why not use `getOrInitDefaultGovernanceConfigs` here?

0xdavinchee replied:
oh derp, good point, the pushed fix resolves this comment and the one above regarding the updatedAt fields...

0xdavinchee committed:
cleanup!

0xdavinchee committed:
leave fields null

0xdavinchee committed:
nullable query

kasparkallas commented on packages/subgraph/src/mappingHelpers.ts:
```diff
@@ -178,6 +207,13 @@ export function getOrInitToken(
     token.isNativeAssetSuperToken = false;
     token.isListed = false;
+
+    // This is unset for the underlying ERC20.
```
Can't leave them undefined?

Also, this is a bit smelly. Having properties on a type that for big portion of the instances don't make sense. Would have probably preferred an explicit TokenGovernance table.

kasparkallas commented on packages/subgraph/src/mappingHelpers.ts:
```diff
@@ -246,6 +282,7 @@ export function getOrInitStream(event: FlowUpdated): Stream {
         stream.streamedUntilUpdatedAt = BigInt.fromI32(0);
         stream.updatedAtTimestamp = currentTimestamp;
         stream.updatedAtBlockNumber = event.block.number;
+        stream.userData = Bytes.empty();
```
Hmm, why not use the `userData` from the event?

0xdavinchee replied:
true

kasparkallas commented on packages/subgraph/src/mappings/superfluidGovernance.ts:
```diff
@@ -61,6 +78,19 @@ export function handleCFAv1LiquidationPeriodChanged(
+            token.liquidationPeriod = event.params.liquidationPeriod;
+            token.save();
+        }
+    }
```
(follow-up comment to "TokenGovernance") most of these code blocks would make more sense like this:
```ts
const governance = getOrInitGovernanceConfig(event.params.superToken.toHexString(), event.block);
governance.liquidationPeriod = event.params.liquidationPeriod;
governance.updatedAtTimestamp = event.block.timestamp;
governance.updatedAtBlock = event.block.number
governance.save();
```
Would not need conditionals in these situations. Also token-specific governance would have timestamps.

0xdavinchee committed:
entity cleanup

- remove DefaultGovernanceConfig, use TokenGovernanceConfig instead
- remove properties from token and instead put it in a reverse mapping with TokenGovernanceConfig
- fix up sdk-core mappings to be in sync

0xdavinchee committed:
Merge branch 'dev' into subgraph-user-data

0xdavinchee committed:
remove gov config queries

0xdavinchee committed:
Merge branch 'subgraph-user-data' of github.com:superfluid-finance/protocol-monorepo into subgraph-user-data

kasparkallas reviewed (approved):
LGTM :+1:

0xdavinchee committed:
lingering BigNumber

0xdavinchee committed:
Merge branch 'dev' into subgraph-user-data

0xdavinchee committed:
Merge branch 'dev' into subgraph-user-data

0xdavinchee committed:
Merge branch 'dev' into subgraph-user-data

hellwolf commented on .github/workflows/call.deploy-subgraph.yml:
```diff
@@ -89,7 +89,10 @@ jobs:
+        # note that this step is only used for generating an arbitrary subgraph.yaml
+        # and addresses.ts file for the yarn codegen step below
+        # the actual addresses will be retrieved during the actual deployment of the subgraph
+        run: ./tasks/prepare-manifest.sh
```
hardcode "matic" here instead.

hellwolf reviewed (approved):
one minor comment, please address.

0xdavinchee committed:
prepare manifest explicit

0xdavinchee committed:
Merge branch 'dev' into subgraph-user-data
